server/src/compilerDiagnostics/diagnostics/SpecifyCompilerVersion.ts

```diff
@@ -0,0 +1,78 @@
+import { TextDocument } from "vscode-languageserver-textdocument";
+import {
+  CodeAction,
+  CodeActionKind,
+  Diagnostic,
+  Range,
+} from "vscode-languageserver/node";
+import { CompilerDiagnostic, ResolveActionsContext } from "../types";
+import { attemptConstrainToFunctionName } from "../conversions/attemptConstrainToFunctionName";
+import { HardhatCompilerError, ServerState } from "../../types";
+
+/**
+ * This diagnostic is shown when no compiler version is specified
+ * i.e. pragma solidity XXXX;
+ *
+ * The suggested quickfix adds the pragma solidity statement, by
+ * extracting the version from hardhat's warning message.
+ */
+export class SpecifyCompilerVersion implements CompilerDiagnostic {
+  public code = "3420";
+  public blocks: string[] = [];
+
+  public fromHardhatCompilerError(
+    document: TextDocument,
+    error: HardhatCompilerError
+  ): Diagnostic {
+    return attemptConstrainToFunctionName(document, error);
```

Comment:
This will work but do an unnecessary check to see if the warning range covers a function (which it won't).

The idea of this function is that it is responsible for transforming the hardhat error (warning/error in reality) into a diagnostic (vscode concept), so it is a natural point to manipulate the range of the hardhat error; which we want to do when we have a whole function error/warning or a whole contract error/warning.

If you are not changing anything about the warning/error and are wanting the default transform you can put:

```javascript
import { passThroughConversion } from "@compilerDiagnostics/conversions/passThroughConversion";
...

public fromHardhatCompilerError(
    document: TextDocument,
    error: HardhatCompilerError
  ): Diagnostic {
    return passThroughConversion(document, error);
  }
```